[#10] 강의 등록 기능 구현

kyeumd opened the pull request:
## 목표

- [x] 강의는 강의명과 강의 자료, 강의 가격과 챕터, 강의 영상들을 업로드한다
- [x] 강의 등록 성공 시 201 Created 반환

## 고려사항

- 영상 포맷
- 대용량 데이터 처리
- Role에 대한 접근 제어

chicori3 committed:
feat(#16) - 공통 예외 코드 작성

chicori3 committed:
feat(#16) - 공통 커스텀 예외 작성

chicori3 committed:
feat(#16) - 공통 응답 객체 작성

chicori3 committed:
feat(#16) - 매개변수 final로 변경

Commit:
feat(#10) - 전체 구조 작성

Commit:
feat(#10) - 강의 등록 기본 구현

Commit:
feat(#10) - 강의 저장 Test 작성

Commit:
feat(#10) - 강의 파일 저장 , 파일 분할 업로드 미구현

Commit:
feat(#10) - 강의 상세 저장

1. 강의 상세 Request Lecture 내부 포함
2. 강의 저장 후 해당 ID를 가지고 강의 상세 저장

Commit:
feat(#10) - 강의 상세 저장 consume 제거

Commit:
feat(#10) - 강의 상세 저장 test mock 객체로 작성

Commit:
feat(#10) - yaml 수정

chicori3 commented on src/main/resources/application.yaml:
```diff
@@ -7,7 +7,7 @@ spring:
       enabled: true
   datasource:
-    url: jdbc:h2:mem:~/local
+    url: jdbc:h2:file:~/local
```
mem이 아니라 file로 다르게 설정하신 이유가 있나요?

kyeumd replied:
jdbc template 같은 경우 테이블을 직접 생성해야 하기 때문에
메모리에 저장하게 되면 매번 테이블을 직접 생성해줘야 하는 문제가 있어 file 로 저장하고 처리하도록 하였습니다.

chicori3 replied:
애플리케이션 실행할 때 미리 작성한 ddl sql 파일 로드해서 테이블 미리 생성하는 방법도 있어요

kyeumd replied:
DDL sql 파일 로드 하는 방법도 한번 고려해볼게요 
데이터 자체도 로컬환경에서 지속적으로 관리 하기 위함도 있었습니다

chicori3 replied:
저게 지금은 사소하지만 나중에 각자 작업할 때 파악이 어려워질 것 같아서 😅

kyeumd replied:
혹시 어떤 관점의 각자 작업 파악을 얘기하시는지 알수있을까요?

chicori3 replied:
단순하게 저는 메모리 DB로 설정해놨고 동겸님은 파일로 설정을 했는데 DB에 에러가 발생하면 코드가 원인인지 설정이 원인인지 파악하기 힘들지 않을까요?

kyeumd replied:
그럴 수 있을것 같습니다. mem을 기본값으로 가져가고 제가 필요하면 file로 사용하고 PR 목록에서는 제외하겠습니다. 
merge 전에 수정해서 commit 치겠습니다!

chicori3 commented on src/main/java/com/flab/infrun/lecture/domain/LectureVideoFile.java:
```diff
@@ -0,0 +1,28 @@
+    private final String url;
+    private final String name;
+
+    public LectureVideoFile(String url, String name) {
```
다른 엔티티 객체들은 private 으로 생성을 제한하고 정적 팩토리 메서드로 생성하게끔 해놓으셨는데 여기는 public으로 설정하신 이유가 있을까요??

kyeumd replied:
private으로 작성되었어야 하는게 맞고 제 의도도 그러했습니다. 수정해야겠네요 지적감사합니다!

f-lab-michael commented on src/main/java/com/flab/infrun/InfrunApplication.java:
```diff
@@ -6,6 +6,7 @@
 @SpringBootApplication
 public class InfrunApplication {
 
+
```
불필요 빈칸

f-lab-michael commented on src/main/java/com/flab/infrun/lecture/application/LectureFacade.java:
```diff
@@ -0,0 +1,23 @@
+        lectureProcessor.uploadFile(lectureVideoFile);
+    }
+
+
```
요기도 불필요하게 2칸이 빈칸이네요. 
전체적으로 스타일과 관련된 체크를 위해서 CheckStyle 같은 플러그인을 도입해보는건 어떨까요?

kyeumd replied:
현재 CodeStyle 을 사용하고 있긴한데 아마 저 사이 공백에 대한 룰이 따로 없어서 그런거 같습니다. 이 부분 한번 확인해보겠습니다!

f-lab-michael commented on src/main/java/com/flab/infrun/lecture/application/LectureProcessor.java:
```diff
@@ -0,0 +1,73 @@
+                    savedLecture.getId())));
+    }
+
+
```
여기도 어떤곳은 메소드 사이가 빈칸이 하나고 어떤곳은 두개

f-lab-michael commented on src/test/java/com/flab/infrun/lecture/infrastructure/persistance/LectureDetailJdbcTemplateRepositoryTest.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
```
참고로 `DataJpaTest` 를 사용하면 JPA 와 관련된 빈들만 사용해서 테스트 할수있습니다.

kyeumd replied:
테스트 관련 DataJpaTest 확인해보겠습니다!

chicori3 commented on src/main/java/com/flab/infrun/lecture/presentation/LectureController.java:
```diff
@@ -0,0 +1,41 @@
+        @RequestPart("lecture") LectureRegisterRequest lecture,
+        @RequestPart("file") List<MultipartFile> lectureVideoFile) {
+
+        log.info("lecture :{}", lecture);
```
log를 남기신 이유가 있나요?

kyeumd replied:
들어오는 request에 대한 정보를 controller 측에서도 파악하기 위해서 작성하였습니다. 추후 이러한 log 관련 추가적인 논의가 필요할 것으로 보입니다.

chicori3 replied:
이 로그는 로컬에서 직접 E2E 테스트를할 때는 유의미할 수 있겠지만 실제 운영서버로 올라갔을 때는 굳이 필요하지 않을 것 같습니다.
어떤 Request가 요청되었는지 우리가 알아야 할 필요가 있을지, 또 파일로 저 로그를 관리해야할 필요가 있을지 의문입니다.

f-lab-michael replied:
요론경우 debug로 로그를 남기고 운영환경에서는 최소 로그 레벨을 info 정도로 바꿔주는 방법을 통해 관리할수있습니다.

kyeumd replied:
debug로 관리하는 방법을 채택하겠습니다! 감사합니다

Commit:
feat(#10) - convention 수정

Commit:
feat(#10) - convention 수정

Commit:
feature(#10) - 강의상세 - 파일 매핑 저장 처리

Commit:
feature(#10) - MyBatisAdapter 추가 작업

Commit:
feature(#10) - test code 수정

Commit:
feat(#10) - mybatis Config build.gradle , yaml

Commit:
feat(#10) - request validation check

Commit:
feat(#10) - 파일명 중복 체크 validation

f-lab-michael commented on src/main/resources/application.yaml:
```diff
@@ -10,4 +10,12 @@ spring:
+    multipart:
+      max-file-size: 50MB
+      max-request-size: 50MB
+mybatis:
```
제가 기억하기로 프로젝트에서JPA를 쓰기로 했던것 같은데 mybatis 를 도입한 이유가 있을까요?

kyeumd replied:
제가 처음에 착각한 부분이 있어서 현재 JPA와 mybatis를 혼용해서 사용하려고 하는데 혹 문제가 되는 상황이 발생할 여지가 있을까요?

f-lab-michael replied:
일반적인 프로젝트 컨벤션의 관점으로 봤을때 jpa repository 의 method name으로 자동생성되는 쿼리나 jpql등으로 컨트롤 하기 어려운 레벨의 쿼리에 있어서 native 쿼리를 직접 사용할때 어쩔수없이 사용하는 경우는 있을것 같은데요, 
지금 진행하시는 방식은 jpa로 쉽게 할수있는 쿼리도 jdbc template을 쓰고 있는것처럼 보여서 불필요하게 2개기술이 사용되는것 같습니다. 이는 프로젝트의 복잡도는 올리기 때문에 가능하면 하나의 기술로 통일시키는게 좋죠.

kyeumd replied:
해당 내용 관련해서 멘토링 시간에 좀 더 여쭤보겠습니다!

f-lab-michael commented:
과거에 리뷰 해놓고 제출을 안했어서 이미 outdated된 comment들이 있으니 관련없는것들은 무시해주세요

Commit:
feat(#10) - log debug 관리

Commit:
Merge branch 'develop' into feature/#10-lecture-register

# Conflicts:
#	src/main/java/com/flab/infrun/common/exception/ErrorCode.java
#	src/main/java/com/flab/infrun/common/response/Response.java
#	src/test/java/com/flab/infrun/InfrunApplicationTests.java

Commit:
feat(#10) - rebase 충돌 수정 // 강의 파일명 중복 체크 unit test 작성

Commit:
feat(#10) - Lecture Column 추가 및 myBatis 오류 수정

Commit:
feat(#10) - Persistence Test MyBatis , Persistance -> Persistence 오타 수정

Commit:
feat(#10) - mybatis test transaction 처리

Commit:
refactor(#10) - id int -> Long type 반환

rusawhale commented on src/main/java/com/flab/infrun/lecture/infrastructure/persistence/LectureVideoFileRepositoryAdapter.java:
```diff
@@ -0,0 +1,27 @@
+        this.lectureVideoFileMyBatisRepository = lectureVideoFileMyBatisRepository;
+    }
+
+    @Override
+    public LectureVideoFile save(LectureVideoFile lectureVideoFile) {
+        return lectureVideoFileMyBatisRepository.save(lectureVideoFile);
+    }
```
file을 디비에 저장하기 보다 object storage 를 활용해 보는게 어떨까요?

rusawhale reviewed (commented):
.

f-lab-michael reviewed (commented):
- jdbcTemplate 을 직접사용하는것과 mybatis 를 사용하는게 둘다 있는것 같은데
어떤기준으로 사용하는건가요?
- 파일을 mysql 에 저장하는것이 아닌 object storage 에 저장하는게 어떨까요?

kyeumd commented:
@f-lab-michael  
1. jdbcTemplate을 활용해서 query 를 작성하였다가 mybatis를 변경하면서 adapter의 구현체를 mybatis로 전부 DI 해준 상태 입니다. 
현재 jdbcTemplate은 직접 사용하지는 않고있습니다!
2. object storage 부분은 내용 찾아보니 매력적인거 같아 채용해보려고 하는데 상용하는 aws s3 같은 상용 api 를 사용하는 것을 말씀하시는게 맞을까요? open source로 되어있는 object storage를 직접 구현해보는 것을 말씀하시는걸까요

chicori3 commented:
> @f-lab-michael
> 
> 1. jdbcTemplate을 활용해서 query 를 작성하였다가 mybatis를 변경하면서 adapter의 구현체를 mybatis로 전부 DI 해준 상태 입니다.
>    현재 jdbcTemplate은 직접 사용하지는 않고있습니다!
> 2. object storage 부분은 내용 찾아보니 매력적인거 같아 채용해보려고 하는데 상용하는 aws s3 같은 상용 api 를 사용하는 것을 말씀하시는게 맞을까요? open source로 되어있는 object storage를 직접 구현해보는 것을 말씀하시는걸까요

이 부분은 인터페이스로 뽑은 다음 지금은 로컬 폴더에 저장하거나 스터빙으로 구현만 해놓고 추후에 클라우드 서비스를 사용해보면서 적용해보면 될 것 같아요

Commit:
refactor(#10) - JPA 도입, Storage Interface 구현
1. JPA 도입
2. myBatis, JdbcTemplate 삭제
3. Storage Save 유형 FileStorage 로 구분
 -> 추후 Object Storage issue 작성

f-lab-michael commented:
@kyeumd 상용 오브젝트 스토리지를 활용하는것을 의미했습니다. 
네이버 클라우드에도 있는것으로 알고 있는데 크레딧 받은걸로 사용하시면 될것 같습니다. 
@chicori3 님 의견처럼 추상화해놓고 이후에 오브젝트 스토리지용 구현체로 바꿔치기 하시면 될것 같습니다.

chicori3 commented on build.gradle:
```diff
@@ -24,18 +24,24 @@ dependencies {
 
     compileOnly 'org.projectlombok:lombok'
 
+    //MyBatis
```
MyBatis를 같이 사용할 여지가 있어서 남기신 건가요??

kyeumd replied:
현재 build.gradle 과 application.yaml 두가지를 충돌을 줄이고자 commit 치지 않고 사용했습니다. 두가지 모두 현행화 해서 다시 올리면 될거같아요

chicori3 commented on src/main/java/com/flab/infrun/common/exception/ErrorCode.java:
```diff
@@ -7,7 +7,8 @@ public enum ErrorCode {
     DUPLICATED_NICKNAME("이미 존재하는 닉네임입니다."),
-    DUPLICATED_EMAIL("이미 존재하는 이메일입니다.");
+    DUPLICATED_EMAIL("이미 존재하는 이메일입니다."),
+    DUPLICATED_FILE_NAME("중복되는 파일명이 존재합니다.");
```
사소하지만 DUPLICATED_FILE_NAME은 MEMBER 하위에 속한다고는 보기 힘든 것 같아요

kyeumd replied:
충돌을 merge처리하는 과정에서 실수가 있었네요 //LECTURE 밑에 들어가도록 수정하겠습니다

chicori3 commented on src/main/java/com/flab/infrun/lecture/application/LectureExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@Slf4j
```
현재 클래스에는 로그를 찍는 부분이 없는데 불필요한 어노테이션이 아닐까요?

kyeumd replied:
불필요한 어노테이션입니다 지적 감사합니다

chicori3 commented on src/main/java/com/flab/infrun/lecture/application/LectureProcessor.java:
```diff
@@ -0,0 +1,90 @@
+        Map<String, LectureFile> result = new HashMap<>();
+        uploadedFile.forEach(
+            file -> result.put(file.getName(), file));
+        return result;
```
이 부분은 스트림을 사용하시겠다면

```
uploadedFiles.stream()
            .collect(Collectors.toMap(LectureFile::getName, Function.identity()));
```

이런 식으로 작성할 수 있을 것 같아요

kyeumd replied:
stream으로 변경해서 map 하는 방식은 생각못했네요 수정해보겠습니다

chicori3 commented on src/main/java/com/flab/infrun/lecture/application/LectureProcessor.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @VisibleForTesting
+    void validateLectureFile(List<MultipartFile> lectureFileList) {
```
테스트를 위한 메서드라고 명시했는데 위의 registerLecture 메서드에서 호출하게 되면 테스트가 아닌 환경에서도 사용되지 않나요?

kyeumd replied:
테스트를 위한 메서드라고 명시한게 아니라 테스트를 하기 위해서 테스트 환경에서 해당 메서드를 테스트하기 위해 visible 되도록 지정해준 것이에요

kyeumd replied:
관련 내용은 제가 VisibleForTesting 을 잘못 알고 사용하고 있었습니다. 짚어주셔서 감사합니다

kyeumd replied:
private 메서드로 사용되어야 하는 것을 테스트를 위해 package-private 으로 작성해야해서 명시적으로 알리기 위해
@VisibleForTesting이 붙게 되었습니다.

chicori3 commented on src/main/java/com/flab/infrun/lecture/domain/Lecture.java:
```diff
@@ -0,0 +1,114 @@
+public class Lecture {
+
+    @Id
+    @GeneratedValue
```
ID 생성 전략을 명시하지 않으면 DEFAULT 값인 AUTO로 설정되는데 이 경우 시퀀스를 지원하지 않는 DB 벤더를 사용한다면
테이블 생성 전략을 사용하게 되요

kyeumd replied:
이부분은 프로젝트에서 MySQL을 사용할것으로 협의해서 일부러 AUTO모드로 둔거긴 한데 해당 내용 조금 더 확인해보겠습니다

kyeumd replied:
MySQL에서 SEQEUNCE를 지원을 안하는군요 확인했습니다. IDENTITY를 사용하도록 변경하였습니다

chicori3 commented on src/main/java/com/flab/infrun/lecture/presentation/request/LectureDetailRequest.java:
```diff
@@ -0,0 +1,20 @@
+
+public record LectureDetailRequest(
+
+    @NotBlank @Size(max = 30)
```
validation이 발생하면 기본 message를 응답으로 보낼텐데 명시적인 메시지를 작성해서 반환하는게 낫지 않을까요?

kyeumd replied:
request에 대한 validation message를 명시적으로 작성하게되면 모든 request에 대해서 일관성 및 통일성이 필요할 것 같습니다. 
그런 것들에 대한 추가 공수보다는 validate가 가지는 기본 message 자체가 이미 통일성과 일관성을 갖추고 있다고 생각했기 떄문에 기본 message를 사용하는 것을 채택했습니다

chicori3 replied:
<img width="832" alt="image" src="https://github.com/f-lab-edu/infrun/assets/40778768/62e973bc-0ec0-4a8c-852b-74c516fa6d31">

메시지를 지정하지 않으면 위와 같은 기본 메시지가 반환되는데
저는 특정 요청을 할 때 요청 데이터의 조건은 "~~이어야 한다"라고 메시지를 직접 작성해서 설명해주는 것이 더 낫다고 생각이 드네요

kyeumd replied:
명시적인것도 좋다고 생각합니다 다만 그렇게 관리를 한다면 직접 문자열을 작성하는 방법이 아닌 메세지 템플릿을 작성해서 관리하는게 좋을것 같습니다.

chicori3 replied:
네 메시지 소스 빈을 만들어 상수로 작서앻두고 사용하는 것도 괜찮을 것 같네요

Commit:
refactor(#10) - 피드백 반영
1. build.gradle, application.yaml 반영
2. ErrorCode 수정
3. 불필요 어노테이션 삭제
4. stream 개선
5. Generated ID default 전략 수정
6. Test code 수정

refactor(#10) - 피드백 반영
1. build.gradle, application.yaml 반영
2. ErrorCode 수정
3. 불필요 어노테이션 삭제
4. stream 개선
5. Generated ID default 전략 수정

kyeumd committed:
Merge branch 'develop' into feature/#10-lecture-register
